### PR TITLE
101-vm-linux-Jupyterhub - new Azure4Students template 

### PR DESCRIPTION
**Pull request by jose-mart**
### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use a parameter for resource locations with the defaultValue set to resourceGroup().location
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values)
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md

- [ ] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog

*
*
*

### CODE REVIEW

File: 101-vm-linux-Jupyterhub/azuredeploy.json [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+    "variables": {
+        "networkInterfaceName": "[concat(parameters('vmName'),'NetInt')]",
+        "virtualMachineName": "[parameters('vmName')]",
+        "publicIpAddressName": "[concat(parameters('vmName'), 'PublicIP' )]",
+        "subnetRef": "[resourceId('Microsoft.Network/virtualNetworks/subnets', parameters('virtualNetworkName'), parameters('subnetName'))]",
+        "nsgId": "[resourceId('Microsoft.Network/networkSecurityGroups', parameters('networkSecurityGroupName'))]",
+        "storageAccountName": "[concat('storage', uniqueString(resourceGroup().id))]",
+        "storageAccountType": "Standard_LRS",
+        "storageAccountKind": "Storage",
+        "vmSize": {
```
Comment (bmoore-msft) [inline 101-vm-linux-Jupyterhub/azuredeploy.json]: this needs to be an explicit user choice - esp given the cpu per user requirement

File: 101-vm-linux-Jupyterhub/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+                        {
+                            "id": "[resourceId('Microsoft.Network/networkInterfaces', variables('networkInterfaceName'))]"
+                        }
+                    ]
+                },
+                "osProfile": {
+                    "computerName": "[variables('virtualMachineName')]",
+                    "adminUsername": "[parameters('adminUsername')]",
+                    "adminPassword": "[parameters('adminPassword')]",
+                    "customData": "[base64(concat('#!/bin/bash \n curl https://raw.githubusercontent.com/jupyterhub/the-littlest-jupyterhub/master/bootstrap/bootstrap.py | sudo python3 - --admin ', parameters('adminUsername'),' >> /var/log/TLJH_install.log 2>&1\n'))]"
```
Comment (bmoore-msft) [inline 101-vm-linux-Jupyterhub/azuredeploy.json]: this script needs to be included in this sample

File: 101-vm-linux-Jupyterhub/azuredeploy.json (hunk 1 above)
Comment (jose-mart) [reply 101-vm-linux-Jupyterhub/azuredeploy.json]: Okay, I think it is solved

File: 101-vm-linux-Jupyterhub/azuredeploy.json (hunk 2 above)
Comment (jose-mart) [reply 101-vm-linux-Jupyterhub/azuredeploy.json]: Added script to install Jupyter

**Comment by jose-mart** [comment]
Hi @bmoore-msft , I added a script file to install Jupyter, but like the file is not yet at the main repository, the CI tests are not passing.

**Comment by leestott** [comment]
@jose-mart I think the script is failing as this point at your repo               "fileUris": ["https://raw.githubusercontent.com/jose-mart/templates/master/101-vm-linux-Jupyterhub/scripts/install.sh"]   not the final destination repo

File: 101-vm-linux-Jupyterhub/azuredeploy.json (hunk 1 above)
Comment (bmoore-msft) [reply 101-vm-linux-Jupyterhub/azuredeploy.json]: I'm still seeing the old pattern

File: 101-vm-linux-Jupyterhub/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+                  "displayName": "Execute my custom script"
+                },
+                "properties": {
+                  "publisher": "Microsoft.Azure.Extensions",
+                  "type": "CustomScript",
+                  "typeHandlerVersion": "2.0",
+                  "autoUpgradeMinorVersion": true,
+                  "protectedSettings": {
+                    "commandToExecute": "[concat('./install.sh ', parameters('adminUsername'))]",
+                    "fileUris": ["https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/101-vm-linux-Jupyterhub/scripts/install.sh"]
```
Comment (bmoore-msft) [inline 101-vm-linux-Jupyterhub/azuredeploy.json]: See [BP guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#deployment-artifacts-nested-templates-scripts) for how to handle artifacts

File: 101-vm-linux-Jupyterhub/scripts/install.sh [hunk 4]
```diff
@@ -0,0 +1,3 @@
+!/bin/bash
+
+curl https://raw.githubusercontent.com/jupyterhub/the-littlest-jupyterhub/master/bootstrap/bootstrap.py | sudo python3 - --admin $1 >> /var/log/TLJH_install.log 2>&1
```
Comment (bmoore-msft) [inline 101-vm-linux-Jupyterhub/scripts/install.sh]: this needs to be included in the sample as well

Comment (bmoore-msft) [reply 101-vm-linux-Jupyterhub/scripts/install.sh]: bootstrap.py needs to be included in this sample

Comment (leestott) [reply 101-vm-linux-Jupyterhub/scripts/install.sh]: @jose-mart  so simply change the install.sh to install bootstrap.py locally so add bootstrap.py to the scripts folder and add the following install.sh

!/bin/bash
bootstrap.py | sudo python3 - --admin $1 >> /var/log/TLJH_install.log 2>&1

Comment (jose-mart) [reply 101-vm-linux-Jupyterhub/scripts/install.sh]: Included

File: 101-vm-linux-Jupyterhub/azuredeploy.json [hunk 5]
```diff
@@ -95,7 +110,27 @@
             "CPU-16GB": "Standard_D4s_v3",
             "GPU-56GB": "Standard_NC6_Promo"
         },
-        "vmName": "[concat(parameters('vmName'),'-', parameters('cpu-gpu'))]"
+        "vmName": "[concat(parameters('vmName'),'-', parameters('cpu-gpu'))]",
+        "dataDisks": {
```
Comment (bmoore-msft) [inline 101-vm-linux-Jupyterhub/azuredeploy.json]: remove this and use implicit disks

File: 101-vm-linux-Jupyterhub/azuredeploy.json [hunk 6]
```diff
@@ -236,14 +271,29 @@
             },
             "kind": "[variables('storageAccountKind')]"
         },
+        {
```
Comment (bmoore-msft) [inline 101-vm-linux-Jupyterhub/azuredeploy.json]: move this and use implicit disks

File: 101-vm-linux-Jupyterhub/azuredeploy.json [hunk 7]
```diff
@@ -261,7 +311,23 @@
                         "offer": "UbuntuServer",
                         "sku": "18.04-LTS",
                         "version": "latest"
-                    }
+                    },
+                    "copy": [
```
Comment (bmoore-msft) [inline 101-vm-linux-Jupyterhub/azuredeploy.json]: remove copy since you're only creating one

File: 101-vm-linux-Jupyterhub/azuredeploy.json [hunk 8]
```diff
@@ -340,6 +297,12 @@
                     "computerName": "[parameters('vmName')]",
                     "adminUsername": "[parameters('adminUsername')]",
                     "adminPassword": "[parameters('adminPassword')]"
+                },
+                "diagnosticsProfile": {
+                    "bootDiagnostics": {
+                        "enabled":  true ,
+                        "storageUri": "[concat(reference(concat('Microsoft.Storage/storageAccounts/', variables('storageAccountName')), '2016-01-01').primaryEndpoints.blob)]"
```
Comment (bmoore-msft) [inline 101-vm-linux-Jupyterhub/azuredeploy.json]: reference(variables('storageAccountName')).primaryEndpoints.blob

**Pull request merged**